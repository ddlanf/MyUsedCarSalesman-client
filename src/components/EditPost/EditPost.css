
.edit-post{
    margin-top: 11%;
}

.edit-post-heading{
    font-size: 7vw;
    text-align: center;
    margin-top: 5%;
    margin-bottom: 6%;
    margin-left: auto;
    margin-right: auto;
    font-family: 'Merriweather Sans', sans-serif;
}

.edit-post-form{
    width: 90%;
    margin-left: auto;
    margin-right: auto;
    margin-bottom: 10%;
    padding-bottom: 2%;
    position: relative;
    background-color: #EAEDEC;
    border: #999090 solid 2px;
    border-radius: 10%;
}

.edit-post-input-box{
    width: 100%;
    margin-left: auto;
    margin-right: auto;
    padding-left: 8%;
    display: flex;
    flex-wrap: wrap;
    flex-direction: row;
    justify-content: stretch;
}


.edit-post-right, .edit-post-left{
    width: 100%;
    min-width: 250px;
    margin: 0;
}


textarea{
    left: 5%;
    position: relative;
    top: -0.2%;
}

.edit-post-label{
    width: 100%;
    display: block;
    font-size: 4vw;
}

.edit-post-input{
    display: block;
    margin-bottom: 3%;
    width: 30%;
    font-size: 4vw;
}

 .edit-post-commission{
    width: 70%;
}

.edit-post-description{
    height: 20vw;
    width: 70%;
 }

 .edit-post-terms-and-cond{
    width: 70%;
    height: 20vw;
 }

.edit-post-image{
    width: 55%;
    height: 5vw;
    font-size: 2vw;
 }

 .edit-post-submit{
    display: block;
    height: 5vw;
    width: 40%;
    height: 10vw;
    max-height: 60px;
    max-width: 300px;
    border-radius: 10%;
    color: black;
    background-color: #999090;
    border: #999090 solid 2px;
    margin-top: 4%;
    margin-bottom: 2%;
    margin-left: auto;
    margin-right: auto;
    font-size: 5vw;
}

.edit-post-submit:hover{
    background-color: #645b5b;
    border: #645b5b solid 2px;
}

.edit-post-error, .edit-post-buffer{
    font-size: 3vw;
    text-align: center;
    border: 1px solid #999090;
    background-color: #DFE32E;
    width: 50%;
    margin-top: 0;
    margin-bottom: 2%;
    margin-left: auto;
    margin-right: auto;
    overflow-x: auto;
}

.edit-post-error{
    color: red;
}

.edit-post-buffer{
    color: black;
}

@media screen and (min-width: 1000px) {

    .edit-post-heading{
        font-size: 60px;
        margin-bottom: 0;
        margin-top: 4%;
    }

    .edit-post-right, .edit-post-left{
        width: 50%;
    }

    .edit-post-input-box{
        width: 100%;
        padding-left: 10%;
    }

    .edit-post-label{
        font-size: 35px;
    }
    
    .edit-post-input{
        width: 50%;
        font-size: 35px;
    }

    .edit-post-commission{
        width: 70%;
    }
    
    .edit-post-description{
        width: 70%;
     }
    
     .edit-post-terms-and-cond{
        width: 70%;
     }
    
    .edit-post-image{
        width: 50%;
     }

     .edit-post-submit{
        margin-top: 2%;
        font-size: 40px;
    }
}