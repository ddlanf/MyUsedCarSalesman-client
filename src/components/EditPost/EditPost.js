import React, { Component } from 'react'
import { withRouter } from 'react-router-dom'
import './EditPost.css'
import PostApiService from '../../services/posts-api-service'
import ImageApiService from '../../services/images-api-service'

class EditPost extends Component {

    _isMounted = false;


    constructor(props){
        super(props);
        this.state = { 
                    make: '',
                    model: '',
                    year: '',
                    mileage: '',
                    description: '',
                    price: '',
                    commission_amount: '',
                    location: '',
                    other_terms_and_conditions : '',
                    image : '',
                    prevPost : {},
                    prevImage : '',
                    clicked: false,
                    buffer: false
        }
    }

    editPost = (ev) =>{

        this.setState({ clicked: true })
        
        ev.preventDefault()
        const { make, model, year, 
            mileage, description, 
            commission_amount, 
            location, price, 
            other_terms_and_conditions, image } = this.state
        
        const tempPost = { make, model, year, 
            mileage, description, 
            commission_amount, 
            location, price, 
            other_terms_and_conditions }
        
        let updatedPost = {}

        for(const key in tempPost){
           
            if(tempPost[key]){
                updatedPost[key] = tempPost[key]
            }
        }

    
        const { history } = this.props
        const { postId } = this.props.match.params

        if(Object.keys(updatedPost).length){
            this.setState({ buffer: true })
            PostApiService.editPost(postId, updatedPost)
                .then(() => {
                    PostApiService.getPosts()
                        .then(posts =>{ 
                            this.context.setPosts(posts) 
                            if(!image){ history.push('/view-posts') }
                        })
                        .catch(res =>{
                            if(this._isMounted){
                                this.setState({ buffer: false, error: res.error })
                            }
                        })
                })
                .catch(res => {
                    history.push('/view-posts') 
                    if(this._isMounted){this.setState({ error: res.error })}
                })
        }
         
        const updatedImage = { src : image } 
    
        if(image){
            ImageApiService.editImage(updatedImage, postId)
                .then(() => {
                    ImageApiService.getImages()
                        .then(images =>{ 
                            this.context.setImages(images) 
                            history.push('/view-posts')
                        })
                        .catch(res =>{
                            if(this._isMounted){
                                this.setState({ buffer: false, error: res.error })
                            }
                        })
                })
                .catch(res => {
                    if(this._isMounted){this.setState({ error: res.error })}
                })
        }

        if(!image && !Object.keys(updatedPost).length){ history.push('/view-posts') }
    }

    handleInputChange = (event) =>{
        const { name } = event.target
        const { value } = event.target
        
        this.setState({
             [name] : value 
        })
    }

    componentDidMount(){

        this._isMounted = true;

        const { postId } = this.props.match.params

        PostApiService.getPost(postId)
            .then(post =>{
               this.setState({ prevPost: post })
            })
            .catch(res => {
                this.setState({ error: res.error })
            })
        ImageApiService.getImage(postId)
            .then(images =>{
               this.setState({ prevImage: images[0].src })
            })
            .catch(res => {
                this.setState({ error: res.error })
            })
    }

    componentWillUnmount() {
        this._isMounted = false;
    }

    render() {
        return (
            <div className="edit-post">
                <section className="edit-post-form-box">
                    <form className="edit-post-form"
                        onSubmit={this.editPost}
                        >
                        <h1 className="edit-post-heading">Edit Post</h1>
                        <div className="edit-post-input-box">
                            <div className="edit-post-left">
                                <label name="edit" className="edit-post-label">Make</label>
                                <input 
                                    type="text" 
                                    className="edit-post-input"
                                    name="make"
                                    defaultValue={this.state.prevPost.make} 
                                    onChange={this.handleInputChange}/>
                                <label name="model" className="edit-post-label">Model</label>
                                <input 
                                    type="text" 
                                    className="edit-post-input"
                                    name="model"
                                    defaultValue={this.state.prevPost.model} 
                                    onChange={this.handleInputChange}/>
                                <label name="year" className="edit-post-label">Year</label>
                                <input 
                                    type="number" 
                                    className="edit-post-input"
                                    name="year"
                                    defaultValue={this.state.prevPost.year} 
                                    onChange={this.handleInputChange}/>
                                <label name="mileage" className="edit-post-label">Mileage</label>
                                <input 
                                    type="number" 
                                    className="edit-post-input"
                                    name="mileage"
                                    defaultValue={this.state.prevPost.mileage} 
                                    onChange={this.handleInputChange}/>
                                <label name="description" className="edit-post-label">Description</label>
                                <textarea 
                                    className="edit-post-description edit-post-input"
                                    name="description"
                                    defaultValue={this.state.prevPost.description} 
                                    onChange={this.handleInputChange}/>
                            </div>
                            <div className="edit-post-right">
                                <label name="price" className="edit-post-label">Price</label>
                                <input 
                                    type="number" 
                                    className="edit-post-input"
                                    name="price"
                                    defaultValue={this.state.prevPost.price} 
                                    onChange={this.handleInputChange}/>
                                <label name="commission" className="edit-post-label">Commision</label>
                                <input 
                                    type="text"
                                    className="edit-post-input edit-post-commission"
                                    name="commission_amount"
                                    defaultValue={this.state.prevPost.commission_amount} 
                                    onChange={this.handleInputChange}/>
                                <label name="location" className="edit-post-label">Location (City, State)</label>
                                <input 
                                    className="edit-post-input edit-post-location"
                                    name="location"
                                    type="text"
                                    defaultValue={this.state.prevPost.location} 
                                    onChange={this.handleInputChange}/>
                                <label name="terms-and-cond" className="edit-post-label">Other Terms and Conditions</label>
                                <textarea 
                                    className="edit-post-terms-and-cond edit-post-input"
                                    name="other_terms_and_conditions"
                                    defaultValue={this.state.prevPost.other_terms_and_conditions} 
                                    onChange={this.handleInputChange}/>
                                <label name="image" htmlFor="img" className="img edit-post-label">Image Link:</label>
                                <input 
                                    className="edit-post-input edit-post-image"
                                    name="image"
                                    type="text"
                                    defaultValue={this.state.prevImage} 
                                    onChange={this.handleInputChange}/>
                            </div>
                        </div>
                        {this.state.buffer ? <p className="make-post-buffer">Uploading please wait...</p> : (this.state.error ? <p className="make-post-error">{this.state.error}</p> : '')}
                        <button disabled={this.state.clicked} id="submit" type="submit" className="edit-post-submit">Submit</button>
                    </form>
                </section> 
            </div>
        )
    }
}

export default withRouter(EditPost)