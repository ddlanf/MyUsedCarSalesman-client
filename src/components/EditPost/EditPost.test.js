import React from 'react';
import ReactDOM from 'react-dom';
import EditPost from './EditPost';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom'

describe('EditPost component', () => {
    
    it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BrowserRouter>
                        <EditPost/>
                    </BrowserRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
    });

    it('renders the UI as expected', () => {
        const tree = renderer
          .create(
                <BrowserRouter>
                    <EditPost/>
                </BrowserRouter>)
          .toJSON();
        expect(tree).toMatchSnapshot();  
      });
});