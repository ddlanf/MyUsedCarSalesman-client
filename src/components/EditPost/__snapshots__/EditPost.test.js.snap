// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPost component renders the UI as expected 1`] = `
<div
  className="edit-post"
>
  <section
    className="edit-post-form-box"
  >
    <form
      className="edit-post-form"
      onSubmit={[Function]}
    >
      <h1
        className="edit-post-heading"
      >
        Edit Post
      </h1>
      <div
        className="edit-post-input-box"
      >
        <div
          className="edit-post-left"
        >
          <label
            className="edit-post-label"
            name="edit"
          >
            Make
          </label>
          <input
            className="edit-post-input"
            name="make"
            onChange={[Function]}
            type="text"
          />
          <label
            className="edit-post-label"
            name="model"
          >
            Model
          </label>
          <input
            className="edit-post-input"
            name="model"
            onChange={[Function]}
            type="text"
          />
          <label
            className="edit-post-label"
            name="year"
          >
            Year
          </label>
          <input
            className="edit-post-input"
            name="year"
            onChange={[Function]}
            type="number"
          />
          <label
            className="edit-post-label"
            name="mileage"
          >
            Mileage
          </label>
          <input
            className="edit-post-input"
            name="mileage"
            onChange={[Function]}
            type="number"
          />
          <label
            className="edit-post-label"
            name="description"
          >
            Description
          </label>
          <textarea
            className="edit-post-description edit-post-input"
            name="description"
            onChange={[Function]}
          />
        </div>
        <div
          className="edit-post-right"
        >
          <label
            className="edit-post-label"
            name="price"
          >
            Price
          </label>
          <input
            className="edit-post-input"
            name="price"
            onChange={[Function]}
            type="number"
          />
          <label
            className="edit-post-label"
            name="commission"
          >
            Commision
          </label>
          <input
            className="edit-post-input edit-post-commission"
            name="commission_amount"
            onChange={[Function]}
            type="text"
          />
          <label
            className="edit-post-label"
            name="location"
          >
            City
          </label>
          <input
            className="edit-post-input edit-post-city"
            name="city"
            onChange={[Function]}
            type="text"
          />
          <label
            className="edit-post-label"
            name="location"
          >
            State
          </label>
          <input
            className="edit-post-input edit-post-state"
            name="state"
            onChange={[Function]}
            type="text"
          />
          <label
            className="edit-post-label"
            name="terms-and-cond"
          >
            Other Terms and Conditions
          </label>
          <textarea
            className="edit-post-terms-and-cond edit-post-input"
            name="other_terms_and_conditions"
            onChange={[Function]}
          />
          <label
            className="img edit-post-label"
            htmlFor="img"
            name="image"
          >
            Image Link:
          </label>
          <input
            className="edit-post-input edit-post-image"
            defaultValue=""
            name="image"
            onChange={[Function]}
            type="text"
          />
        </div>
      </div>
      
      <button
        className="edit-post-submit"
        disabled={false}
        id="submit"
        style={
          Object {
            "backgroundColor": "#999090",
          }
        }
        type="submit"
      >
        Submit
      </button>
    </form>
  </section>
</div>
`;
